pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,351 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2020 Datadog, Inc.
+
+package listeners
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/persistentcache"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/soniah/gosnmp"
+)
+
+func init() {
+	Register("snmp", NewSNMPListener)
+}
+
+// SNMPListener implements SNMP discovery
+type SNMPListener struct {
+	sync.RWMutex
+	newService chan<- Service
+	delService chan<- Service
+	stop       chan bool
+	config     util.SNMPListenerConfig
+	services   map[string]Service
+	devices    map[string][]string
+}
+
+// SNMPService implements and store results from the Service interface for the SNMP listener
+type SNMPService struct {
+	adIdentifier string
+	entityID     string
+	deviceIP     string
+	creationTime integration.CreationTime
+}
+
+// Make sure SNMPService implements the Service interface
+var _ Service = &SNMPService{}
+
+type snmpSubnet struct {
+	adIdentifier  string
+	config        util.SNMPConfig
+	defaultParams *gosnmp.GoSNMP
+	startingIP    net.IP
+	currentIP     net.IP
+	network       net.IPNet
+	cacheKey      string
+}
+
+// NewSNMPListener creates a SNMPListener
+func NewSNMPListener() (ServiceListener, error) {
+	var snmpConfig util.SNMPListenerConfig
+	err := config.Datadog.UnmarshalKey("snmp_listener", &snmpConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &SNMPListener{
+		services: map[string]Service{},
+		devices:  map[string][]string{},
+		stop:     make(chan bool),
+		config:   snmpConfig,
+	}, nil
+}
+
+// Listen periodically refreshes devices
+func (l *SNMPListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go l.checkDevices()
+}
+
+func (l *SNMPListener) loadCache(config util.SNMPConfig, adIdentifier string, cacheKey string) {
+	cacheValue, err := persistentcache.Read(cacheKey)
+	if err != nil {
+		log.Errorf("Couldn't read cache for %s: %s", cacheKey, err)
+		return
+	}
+	if cacheValue == "" {
+		return
+	}
+	var devices []net.IP
+	err = json.Unmarshal([]byte(cacheValue), &devices)
+	if err != nil {
+		log.Errorf("Couldn't unmarshal cache for %s: %s", cacheKey, err)
+		return
+	}
+	for _, deviceIP := range devices {
+		entityID := config.Digest(deviceIP.String())
+		l.createService(deviceIP.String(), adIdentifier, entityID)
+	}
+}
+
+func writeCache(cacheKey string, devices []string) {
+	cacheValue, err := json.Marshal(devices)
+	if err != nil {
+		log.Errorf("Couldn't marshal cache: %s", err)
+		return
+	}
+	err = persistentcache.Write(cacheKey, string(cacheValue))
+	if err != nil {
+		log.Errorf("Couldn't write cache: %s", err)
+	}
```

Comment:
```suggestion
	if err = persistentcache.Write(cacheKey, string(cacheValue)); err != nil {
		log.Errorf("Couldn't write cache: %s", err)
	}
```
but I'd rather return the error itself
`return persistentcache.Write(cacheKey, string(cacheValue))`